In [ ]:
from sklearn import datasets
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import ParameterGrid
from sklearn.model_selection import GridSearchCV
import time
from sklearn.model_selection import RandomizedSearchCV



> п.2 - п.3



In [ ]:
iris = datasets.load_iris()
print(iris.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])




> п.4


In [ ]:
print(iris['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========



> п.5



В наборе данных iris значения признаков и меток классов хранятся в следующих ключах:

* data: содержит признаки для каждого образца.


* target: содержит целевые метки классов для каждого образца.

In [ ]:
X = iris.data
print(X)
y = iris.target
print(y)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

**X** содержит признаки для каждого образца, а **y** - целевые метки классов для каждого образца.



> п.6



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print("Размеры обучающей выборки:", X_train.shape, y_train.shape)
print("Размеры тестовой выборки:", X_test.shape, y_test.shape)

Размеры обучающей выборки: (105, 4) (105,)
Размеры тестовой выборки: (45, 4) (45,)




> п.7



In [ ]:
clf = SGDClassifier()
print(clf.get_params())

{'alpha': 0.0001, 'average': False, 'class_weight': None, 'early_stopping': False, 'epsilon': 0.1, 'eta0': 0.0, 'fit_intercept': True, 'l1_ratio': 0.15, 'learning_rate': 'optimal', 'loss': 'hinge', 'max_iter': 1000, 'n_iter_no_change': 5, 'n_jobs': None, 'penalty': 'l2', 'power_t': 0.5, 'random_state': None, 'shuffle': True, 'tol': 0.001, 'validation_fraction': 0.1, 'verbose': 0, 'warm_start': False}






> п.8

Создать  сетку  параметров,  включающую  4  вида  функции  потерь,  два  типа регуляризаторов, 5 значений коэффициента регуляризации от 0.0001 до 0.001, и число итераций от 5 до 10 с шагом 1.





In [ ]:
param_grid = {
    'loss': ['hinge', 'log_loss', 'modified_huber', 'squared_hinge'],
    'penalty': ['l1', 'l2'],
    'alpha': [0.0001, 0.0005, 0.0007, 0.0009, 0.001],
    'n_iter_no_change': [5, 6, 7, 8, 9, 10]
}

grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy')

start_time = time.time()

grid_search.fit(X_train, y_train)

end_time = time.time()
print(f"Время выполнения: {end_time - start_time:.2f} секунд")

Время выполнения: 5.46 секунд


In [ ]:
print("Лучший классификатор: ", grid_search.best_estimator_)
print("Лучший набор параметров: ", grid_search.best_params_)
print(f"Оценка лучшего классификатора: {grid_search.best_score_:.3f}")

Лучший классификатор:  SGDClassifier(alpha=0.0005, loss='log_loss', n_iter_no_change=8, penalty='l1')
Лучший набор параметров:  {'alpha': 0.0005, 'loss': 'log_loss', 'n_iter_no_change': 8, 'penalty': 'l1'}
Оценка лучшего классификатора: 0.971


In [ ]:
results = grid_search.cv_results_
last_10_scores = results['mean_test_score'][-10:]
print("Оценки алгоритма по последним 10 наборам параметров", last_10_scores)

Оценки алгоритма по последним 10 наборам параметров [0.93333333 0.86666667 0.95238095 0.85714286 0.93333333 0.82857143
 0.95238095 0.73333333 0.95238095 0.96190476]


Из результатов видно, что лучший классификатор достиг точности 0.971 на обучающей выборке, что является довольно высоким значением. Набор лучших параметров включает значение коэффициента регуляризации alpha=0.0005, функцию потерь log_loss, показатель остановки при отсутствии улучшения n_iter_no_change=8 и тип регуляризации L1.

Оценки алгоритма по последним 10 наборам параметров показывают, что качество классификации сильно колеблется, с оценками в диапазоне от 0.733 до 0.961. Это может быть связано с тем, что на некоторых комбинациях параметров классификатор плохо справляется с задачей.


**Параметры функции потерь и регуляризатора:** в данном случае, регуляризатор "l1" оказался более эффективным, чем "l2".

**Коэффициент регуляризации:** в данном случае, наилучшие результаты были достигнуты при коэффициенте 0.0005, что может указывать на то, что данные были достаточно сложными для модели и необходимо было использовать сильную регуляризацию.



In [ ]:
param_distributions = {
    'loss': ['hinge', 'log_loss', 'modified_huber', 'squared_hinge'],
    'penalty': ['l1', 'l2'],
    'alpha': [0.0001, 0.0005, 0.0007, 0.0009, 0.001],
    'n_iter_no_change': [5, 6, 7, 8, 9, 10]
}

random_search = RandomizedSearchCV(clf, param_distributions, n_iter=50, cv=5, scoring='accuracy', random_state=42)

start_time = time.time()

random_search.fit(X_train, y_train)

end_time = time.time()
print(f"Время выполнения: {end_time - start_time:.2f} секунд")

print("Лучший классификатор: ", random_search.best_estimator_)
print("Лучший набор параметров: ", random_search.best_params_)
print(f"Оценка лучшего классификатора: {random_search.best_score_:.3f}")

results = random_search.cv_results_
last_10_scores = results['mean_test_score'][-10:]
print("Оценки алгоритма по последним 10 наборам параметров", last_10_scores)

Время выполнения: 1.62 секунд
Лучший классификатор:  SGDClassifier(alpha=0.0007, n_iter_no_change=9, penalty='l1')
Лучший набор параметров:  {'penalty': 'l1', 'n_iter_no_change': 9, 'loss': 'hinge', 'alpha': 0.0007}
Оценка лучшего классификатора: 0.962
Оценки алгоритма по последним 10 наборам параметров [0.86666667 0.91428571 0.80952381 0.94285714 0.9047619  0.93333333
 0.87619048 0.84761905 0.88571429 0.92380952]


Различия в наборах параметров могут быть объяснены случайностью в выборе параметров при случайном поиске, тогда как GridSearchCV перебирает все комбинации параметров из заданных сеток.

В результате, случайный поиск обычно работает быстрее, но может не дать наилучших результатов в сравнении с GridSearchCV.

В данном случае, время выполнения для случайного поиска составляет 1.62 секунды, что быстрее, чем время выполнения для GridSearchCV, которое составляет 4.61 секунды. Это может быть важно при работе с большими объемами данных и большим количеством параметров. Однако, оценка лучшего классификатора, полученная с помощью GridSearchCV (0.971), выше, чем оценка, полученная с помощью случайного поиска (0.962).


Оценки алгоритма по последним 10 наборам параметров для GridSearchCV и случайного поиска различаются. Это может указывать на различия в степени оптимизации параметров между двумя методами.

**Рекомендация**:  использовать случайный поиск для поиска оптимальных параметров, когда необходимо работать с большими объемами данных или большим количеством параметров. Однако, если нужно найти наилучшую модель с высокой точностью, GridSearchCV может быть более эффективным, но будет требовать немного больше времени.